common/djangoapps/student/models.py

```diff
@@ -541,6 +542,16 @@ def level_of_education_display(self):
         if self.level_of_education:
             return self.__enumerable_to_display(self.LEVEL_OF_EDUCATION_CHOICES, self.level_of_education)
 
+    @property
+    def display_name_with_default(self):
+        set_custom_metric('missing_user_profile_display_name', False)
+        if not self.display_name:
+            set_custom_metric('missing_user_profile_display_name', True)
+            self.display_name = self.user.username
+            self.save()
```

Comment:
@robrap @feanil  Edits during reads is not ideal generally speaking - causes issues down the line.  Why do we believe we need this?

Can we do a database search instead and ensure all non-service-users have a profile?  Alternatively, can we ensure this happens in a management command?  (CQRS)

Reply:
This is meant to be a temporary change, the idea being that we can fall-back and auto update display name for users that are active right away. And then use a management command after the fact to backfill everyone else.

Once the backfile is complete we make the field non-nullable and can remove this fallback logic.

Reply:
Thinking about this a bit more, the set/save doesn’t really provide much benefit, so if we are concerned with any potential risk, we can remove it. We want to leave the metric so we can eventually determine if there are any missing values.

Reply:
Thanks for the additional context on the intended purpose of this change. The following is a moot point since we're reconsidering this approach anyway. However, I thought I'd share in case it's useful in similar circumstances in the future.

In addition to concerns of inadvertent downstream issues due to not following CSRS, here were other thoughts that went through my head:

* We would be modifying users' database entries in GET requests, which violates internet standards and makes us vulnerable to CSRF issues (viz "unsafe GETs"). While the CSRF issue may be innocuous in this case, it's still an added violation. In fact, I do hope we can finally run our CSRF middleware checker soon - now that DOP is out the door. (DOP had a CSRF violation that prevented us from enabling the middleware, IIRC.)

* We would be emitting user-change events at random unrelated times. We need to access a user's display-name at all various times - whether when the user lands on an IDA, an MFE, login, or when the system generates a report, etc.

This pattern of lazy-initialization currently also happens for cohort-assignment of enrollees. I'm not a proponent of it - it has similar issues to the ones listed on this thread. For example, a user's cohort gets assigned at all various times - including when the user accesses the course outline or when an admin generates a grade report. Assigning the user to a cohort at the time of enrollment would have provided a consistent point of assignment and fewer issues to debug.

Given this thread, I'm wondering whether we should introduce the following [manifesto item](https://openedx.atlassian.net/wiki/spaces/AC/pages/1074397222/Architecture+Manifesto+WIP):

`CQRS beats Lazy Initialization`